### PR TITLE
Parallel Algorithm Classification Tutorial

### PR DESCRIPTION
**Pull request by michaelmckinsey1**
the `parallel-sorting` dataset is 143.4 MB. The dataset is cleansed for PII.

- [08A_composing_parallel_sorting_data.ipynb](https://github.com/LLNL/thicket-tutorial/blob/parallel-sorting/notebooks/08A_composing_parallel_sorting_data.ipynb) - starts from reading the files and ends at pickling the data for modeling.
- [08B_modeling_parallel_sorting_data.ipynb](https://github.com/LLNL/thicket-tutorial/blob/parallel-sorting/notebooks/08B_modeling_parallel_sorting_data.ipynb) - is the modeling

### CODE REVIEW

**Comment by michaelmckinsey1** [comment]
The notebooks are ran right now for review but will need to be cleared before merging

File: notebooks/08A_composing_parallel_sorting_data.ipynb
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb:1]: It would be awesome to have a simplified version of table 6 from your paper in the first markdown cell. Basically showing an overview of the data you're about to load, the 5 alogrithms and how many samples of each there are. I would also mention that the algorithms were profiled using caliper - and if you have a link to the annotated algorithm source code you can add that would be even better.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 1]
```diff
@@ -88,7 +98,21 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 2. Modify and Filter Metadata Values"
+    "## 3. Modify and Filter Metadata Values\n",
+    "\n",
+    "Since the dataset we are using is a compilation from many different users, there are various errors in the metadata annotations which we can fix using Thicket. We have defined two dictionaries from manual analysis of the data to achieve this:\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I wouldn't call these errors, they are just inconsistencies in naming that you have to clean up. I'd probably just say "labeling inconsistencies" or something instead of "errors"

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 2]
```diff
@@ -88,7 +98,21 @@
...
+    "- `META_FIX_DICT` is used to enforce consistency in the metadata by replacing values.\n",
+    "- `META_WHITELIST_DICT` is used to select the metadata parameters we are looking for from the experiments.\n",
+    "\n",
+    "The metadata we reference are the experiment parameters and important identifying metadata:\n",
+    "\n",
+    "- `InputType` - The type of sortedness of the input array.\n",
+    "- `Datatype` - The datatype of the values in the input array.\n",
+    "- `num_procs` - Number of parallel processes.\n",
+    "- `InputSize` - Size of the input array.\n",
+    "- `Algorithm` - The name of the parallel sorting algorithm.\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I would just change the wording here to be more ML focused. Say you are whitelisting only the profiles that contain the correct column values, that will eventually be "input features" or "labels". And maybe call out that the Algorithm name is the label column specifically.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 3]
```diff
@@ -226,7 +236,9 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 2B. Filter Metadata Values from Whitelist"
+    "### 3B. Filter Metadata Values from Whitelist\n",
+    "\n",
+    "We use the `Thicket.filter_metadata()` function to filter any values that are not contained in our metadata whitelist."
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: Again here, just include the purpose of this step by adding "to include only profiles with the desired input features and labels" or something like that.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 4]
```diff
@@ -238,24 +250,26 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 2C. Filter Duplicate Metadata Values\n",
+    "### 3C. Filter Duplicate Metadata Values\n",
+    "\n",
+    "Duplicate values across all of our experiment parameters indicates that one profile has incorrect metadata, since all of the profiles are single-trial. This is typically user error (metadata is manually annotated in Adiak).\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: This is kind of confusing. Do you remove all entries that have duplication? Also do you expect people reading this to know what Adiak is without introduction? Maybe can reword the second sentence to "If we find duplicates of any profile, we remove them all as we cannot determine which has the correct metadata. These occurrences are a result of incorrect manual annotation before generating the profiles."

Of course this raises the question of was any of the data validated if it's all manually annotated? Hopefully we can trust that people did it right in all other cases...

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 5]
```diff
@@ -281,23 +295,27 @@
...
-    "print(f\"Total files after removing duplicates: {len(tk.profile)}\")"
+    "print(f\"Total profiles after removing duplicates: {len(tk.profile)}\")"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 3. Create Features"
+    "## 4. Create Features\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: We aren't really "creating features" maybe say you are "Creating Input Feature Dataset" or "Selecting Features"

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 6]
```diff
@@ -281,23 +295,27 @@
...
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 3A. Query the Call Tree\n",
+    "### 4A. Query the Call Tree and Re-compose the Thicket\n",
     "\n",
-    "To get performance metrics per node."
+    "Here we select only the nodes we want to compare for our modeling using `Thicket.query()` and `Thicket.concat_thickets()`.\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: What does this query do in plain English from the perspective of how it changes the dataset? What are the values in the "nodes" list below and how do they relate to the data? I think you need another sentence after this that says "In this case, we keep only profiles that satisfy X requirement, then we remove profiles that contain data in the "libraries" and "cmdline" columns." (Honestly, I don't understand what you are doing in this code cell with the query...I think more explanation is needed).

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 7]
```diff
@@ -321,59 +340,63 @@
...
    "metadata": {},
-   "outputs": [],
    "source": [
-    "# Drop duplicate profiles from concat_thickets\n",
-    "# Can't pass these cols in the check or error will be thrown. Won't change the outcome of the check\n",
-    "unhashable_cols = [\"libraries\", \"cmdline\"]\n",
-    "tk.metadata = tk.metadata.drop_duplicates(subset=[col for col in tk.metadata.columns if col not in unhashable_cols])"
+    "### 4B. Remove Profiles not Containing All Nodes\n",
+    "\n",
+    "Because our features will use all of the nodes, we remove profiles that do not have data for all nodes using `Thicket.filter_profile`."
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I'm confused again here, maybe it would help if you introduce what "comp_small" and "comp_large" and all these things mean earlier? I think of a node as one row, so I see you would want every node to have data for every feature that you use as input...is that what you are checking here? How do the comp_* labels relate to each algorithm profile?

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 8]
```diff
@@ -321,59 +340,63 @@
...
-    "tk = tk.filter_profile(profile_truth)"
+    "tk = tk.filter_profile(profile_truth)\n",
+    "print(f\"Total profiles that contain all data: {len(tk.profile)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4C. Compute Features from Performance Data\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: Maybe change to "Compute Additional Features from Performance Data"

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 9]
```diff
@@ -385,94 +408,100 @@
...
    "source": [
-    "# Compute comp/comm\n",
-    "tk.add_root_node(attrs={\"name\": \"comp/comm\"})\n",
-    "compcomm = [node for node in tk.graph.traverse() if node.frame[\"name\"] == \"comp/comm\"][0]\n",
-    "tdf = tk.dataframe.loc[node_objects[\"comp\"], metric_cols].div(tk.dataframe.loc[node_objects[\"comm\"], metric_cols])\n",
-    "for prof in tdf.index:\n",
-    "    tk.dataframe.loc[(compcomm, prof), metric_cols] = tdf.loc[prof]"
+    "### 4D: Define Our Features Using Pandas Slices\n",
+    "\n",
+    "To subselect the performance data we care about we use a slice generated by either `perf_idx()` or `presence_idx()` (they are functions because the node objects can change `id`'s after certain Thicket operations). We use the `Thicket.get_node()` function to select node objects.\n",
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: You describe what the code does, but I think you need more high-level explanation here. Something like "The pef_idx() function returns only data that satisfies X, and the presence_idx() function returns only data that satisfies Y condition." Basically the "why" is missing from this section.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 10]
```diff
@@ -385,94 +408,100 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 3C. Filter Features with NaN Values"
+    "### 4E. Filter Features with NaN Values\n",
+    "\n",
+    "Here we check one last time for any missing data points in any of the profiles for each of the slices we just defined. `any_nan_rows_series` will be a series of boolean values for each profile that will be `True` if there are any missing data points. We use the `Thicket.filter_profile()` function once again to filter out the profiles with missing data points."
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I'm not sure I'm following why there might still be NaNs...an example might be useful here.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 11]
```diff
@@ -385,94 +408,100 @@
...
+    "tk = tk.filter_profile([p for p in tk.profile if p not in nan_profs])\n",
     "print(f\"Total profiles after dropping NaNs: {len(tk.profile)}\")"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 4. Remove Anomalies "
+    "## 5. Remove Anomalies "
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I like the description here, it's good.

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 12]
```diff
@@ -483,83 +512,49 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# 5. Write Model Data"
+    "## 6. Write Model Data\n",
+    "\n",
+    "Lastly we shuffle the data using `pandas.DataFrame.sample()` and pickle the Thicket object, which we will use to pick back up in the next notebook, part B, where we will create classification models using the performance data."
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: You could add more context here, shuffling helps to randomize the train and test sampling later. Could add that pickling is helpful for writing and reading large datasets to/from file, though maybe people would be expected to know that, not sure.

File: notebooks/08B_modeling_parallel_sorting_data.ipynb
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb:1]: I would just add to the description at the top that if they have not stepped through notebook 08A they should start there (and name the notebook so they can find it)

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 13]
```diff
@@ -27,6 +46,16 @@
     "import thicket as th"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Define Modeling Helper Functions\n",
+    "- `prep_data` performs standardization scaling on the numerical columns and converts categorical columns to the correct format for modeling.\n",
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: Could add more explanation, that standard scaling normalizes the data to reduce the effect of outliers skewing the training. Also that categorical columns need to be "one-hot encoded" basically turned into boolean columns so that they are numerical data (if that is indeed what you are doing).

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 14]
```diff
@@ -27,6 +46,16 @@
...
+    "- `split_X_y` splits a dataset into input data and labels.\n",
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: just specify input features (X) and labels (y)

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 15]
```diff
@@ -127,6 +168,14 @@
     "tk = th.Thicket.from_pickle(\"thicket-modeldata.pkl\")"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3B. Concatenate Features\n",
+    "To match the expected format of the Scikit-learn models we will use, concatenate the features together such that each row in the `model_data` DataFrame is a data sample (profile). We can achieve this desired format by using `pd.DataFrame.unstack()` to unstack the node index, and using `pd.concat()` to concatenate the results."
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: This will probably make more sense once 8A is updated with more explanation, but you could add a little more about the structure of the data here and how it changes or include an example.

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 16]
```diff
@@ -143,6 +192,13 @@
     ")"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3C. Define Categorical and Numerical Columns"
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: Add description to 3C: "We manually define a set of the categorical columns and then subtract them from a set of all of the columns to get the list of numerical columns"

I am hacking these comments a bit since github doesn't show changes very well...but for 3D (convert numeric to categorical) can you add a description of why you might do that when you just went the other direction?

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 17]
```diff
@@ -156,6 +212,13 @@
     "numerical_columns = list(set(model_data.columns) - set(categorical_columns + [\"Algorithm\"]))"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3D. Define Dictionary of Classifiers"
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: Add description: "Here we create a dictionary of each of the machine learning models we want to use to classify our algorithm dataset so that we can use them in our loop in step 4." Or something like that

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 18]
```diff
@@ -264,9 +341,17 @@
     "        model_results = pd.concat([model_results, df])"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Visualize Model Performance\n",
+    "With the `_plot_bars()` function, we can visualize the per-fold accuracy of each classifier for each model performance statistic. We notice that although performance varies slightly across folds, the tree models perform the best on performance data."
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: It actually looks like the decision tree is the worst in every single fold and random forest and SVM vary between which is best, should be more specific here.

**Review by dyokelson — COMMENTED** [review]
Overall looking good, mostly needs more high level description text in places. I had to hack the comments in some places because github only lets me view one commit per single file view and was trying to load all the data files...so it was difficult to just review the notebooks with all changes. Let me know if you have any questions about any of the comments in case placement doesn't make sense. They say "outdated" because they are on older commits because the newest commit wouldn't let me comment on lines that weren't specific to that commit...argh github...

File: notebooks/08A_composing_parallel_sorting_data.ipynb
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb:1]: 65af493f 
Agree with all, except we don't have a public dataset for the source code.

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 1 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 33d74fc6

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 2 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 9da50597
Agree except none of these will be used in modeling (other than `Algorithm`) they are only for processing.

Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: abf75e27
Adds context

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 3 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: fb3bf678

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 4 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: e038cd3d
Yes this is removing all profiles that have duplicated metadata, which is not a lot. For example we remove one set of 27 profiles and one set of 26. Since this is single trial data, only 1/27 and 1/26 profiles have correct metadata, so it is not a huge loss to lose 2 correctly annotated profiles.

> Of course this raises the question of was any of the data validated if it's all manually annotated? Hopefully we can trust that people did it right in all other 

The dataset is not 100% validated, which would require manually inspecting each implementation for correctness. Although we do take a stab at this with anomaly removal e.g. if someone annotated a value as 1024 procs but ran with 2 procs, it would be removed as a slow outlier, but this does not solve the problem completely.

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 5 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: bb73193e

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 6 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: This is the hardest part to explain. Here's an example:

This is the tree before step 4
```
0.001 correctness_check
0.000 data_init
nan main
├─ nan MPI_Barrier
├─ nan comm
│  ├─ nan MPI_Barrier
│  ├─ nan comm_barrier
│  ├─ nan comm_large
│  │  ├─ nan MPI_Alltoall
│  │  ├─ nan MPI_Alltoallv
│  │  ├─ nan MPI_Bcast
│  │  ├─ nan MPI_Gather
│  │  ├─ nan MPI_Gatherv
│  │  ├─ nan MPI_Isend
│  │  ├─ nan MPI_Recv
│  │  ├─ nan MPI_Scatter
│  │  ├─ nan MPI_Send
│  │  ├─ nan MPI_Sendrecv
│  │  ├─ nan comm_small
│  │  ├─ nan mpi_gather
│  │  └─ nan mpi_scatter
│  ├─ nan comm_large_gather
│  ├─ nan comm_large_scatter
│  └─ nan comm_small
│     ├─ nan MPI_Allgather
│     ├─ nan MPI_Allreduce
│     ├─ nan MPI_Alltoall
│     ├─ nan MPI_Bcast
│     ├─ nan MPI_Gather
│     ├─ nan MPI_Isend
│     ├─ nan MPI_Recv
│     ├─ nan MPI_Reduce
│     ├─ nan MPI_Send
│     └─ nan MPI_Sendrecv
├─ nan comm_large
├─ nan comm_small
├─ nan comp
│  ├─ nan comm
│  │  ├─ nan comm_large
│  │  │  └─ nan Gatherv
│  │  ├─ nan comm_small
│  │  │  ├─ nan Allgather
│  │  │  └─ nan Gather
│  │  ├─ nan comp_large
│  │  └─ nan comp_small
│  ├─ nan comp_large
│  │  ├─ nan comm
│  │  │  └─ nan comm_large
│  │  ├─ nan comp_small
│  │  └─ nan local_sort
│  └─ nan comp_small
├─ nan comp_large
├─ nan comp_small
├─ nan correct_check
├─ nan correctness_check
├─ nan data_init
├─ nan data_initialization
├─ nan data_validation
├─ nan gatherv
├─ nan main
│  ├─ nan comm
│  │  ├─ nan comm_large
│  │  └─ nan comm_small
│  ├─ nan comp
│  │  ├─ nan comp_large
│  │  └─ nan comp_small
│  ├─ nan correct_check
│  ├─ nan data_init
│  └─ nan mpibarrier
├─ nan mpi_recv
├─ nan mpi_send
├─ nan mpibarrier
├─ nan scatter
├─ nan scatterv
└─ nan whole_computation
   ├─ nan check_correctness
   ├─ nan comm
   │  ├─ nan comm_gather
   │  ├─ nan comm_large
   │  │  ├─ nan MPI_Gather
   │  │  ├─ nan MPI_Recv
   │  │  ├─ nan MPI_Scatter
   │  │  └─ nan MPI_Send
   │  ├─ nan comm_scatter
   │  └─ nan comp_large
   │     └─ nan comm_large
   ├─ nan comp
   │  ├─ nan comp_large
   │  │  └─ nan comm_large
   │  └─ nan comp_small
   └─ nan data_init
0.016 whole_comp_time
├─ 0.011 comm
├─ nan comm_large
├─ nan comm_small
├─ 0.005 comp
├─ nan comp_large
└─ nan comp_small
nan whole_computation
├─ nan comm
│  ├─ nan comm_large
│  └─ nan comm_small
├─ nan comp
│  ├─ nan comp_large
│  └─ nan comp_small
├─ nan correctness_check
└─ nan data_init
```
This is the tree after 4A
```
0.072 comm
0.072 comm_large
nan comm_small
0.062 comp
0.062 comp_large
nan comp_small
```
Clearly, all of the implementations did not perfectly match the generalized tree. Therefore, by composing all of the data (and graphs), the tree becomes really messy. We don't care about most of the nodes in this tree, but we do want to de-duplicate nodes that exist at different levels to make accessing the data more simple.

As a result of creating the generalized tree and enforcing annotations from it, the parent/child relationships in the graph don't matter anymore. So in a sense we don't even need the graph we just need to select the nodes from the generalized tree that we are going to use.

Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: f6c13c3c should help with the clarity issue since the query and composition do not have to be explained in the same cell and are different operations.

Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 27f2074f I took a stab at adding context to the query, but I am not sure exactly how to explain this to a user.

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 7 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 50e25978 I reworded this cell because it isn't necessary to think about nodes here but really we are just removing profiles with missing data.

There is still the issue of whether or not we want to explain the generalized calltree in this tutorial or not.

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 8 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: b92c91d5

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 9 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 520c08ce - these are basically macros but I don't know if they are more confusing or beneficial. Would it be easier to omit this section entirely and just use the full slice every time?

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 10 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 3d1ceb7e - While 4C and 4F are similar, 4C is removing profiles with missing regions and 4F is removing profiles with missing metrics. 4F is simply easier to perform after we define the slices.

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 12 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 2acca657

File: notebooks/08B_modeling_parallel_sorting_data.ipynb
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb:1]: 7440090c - added it twice

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 13 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 69001296

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 14 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 33329a30

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 15 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 06b84ac3

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 16 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 31d96731

Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 090cfc31 3D is a technique described in the journal paper

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 17 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 2647165d

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 18 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 7e92e1a0

File: notebooks/08A_composing_parallel_sorting_data.ipynb [hunk 19]
```diff
@@ -0,0 +1,714 @@
...
+    "In this section, we structure the performance data where each column is a feature, and each row is a feature vector for one performance profile, which is necessary for modeling"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4A. Query the Call Tree\n",
+    "\n",
+    "Here we select only the nodes we want to compare for our modeling using `Thicket.query()`. Viewing the `Thicket.tree()` at this point will show many nodes which are not relevant to our analysis. Query will also restructure the tree, such that it is easier to select performance data."
```
Comment (dyokelson) [inline notebooks/08A_composing_parallel_sorting_data.ipynb]: I think this section still needs a little more clarification. Which might be useful by giving an example of what the source code/structure looks like since we can't look at the repo or at a dataframe printout. For example you could say we filter to keep only the nodes (functions) that are the implementations of each algorithm type. E.g. our algorithm code structure might look like this, for a single profile, where the metadata tag contains the algorithm type, and each of comp, comp_large etc are implementations of that sorting algorithm with different input parameters: 

- main
-- comp (parameters x)
-- comp_large (parameters y)

Filtering for just the algorithm implementations in the performance data removes any functions that aren't sorting algorithms.

or an example table might help (I can't figure out how to do multirow in md right now, but you could mock something up another way maybe):
|node| profile| time|
--- | --- | --- | 
|comp|profile1|2|
|comp|profile2|3|
|comp_large|profile1|2|
|comp_large|profile2|4|

File: notebooks/08B_modeling_parallel_sorting_data.ipynb [hunk 20]
```diff
@@ -0,0 +1,561 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 3C. Define Categorical and Numerical Columns\n",
+    "\n",
+    "We manually define the two categorical columns and then define the numerical columns, which are by definition the remaining columns in the data."
```
Comment (dyokelson) [inline notebooks/08B_modeling_parallel_sorting_data.ipynb]: Maybe here you can just remind them that "Node presence" was a derived feature from the first notebook in this description. I'm still not quite understanding what the feature means, but that is probably described in your paper?

**Review by dyokelson — COMMENTED** [review]
Looks good, just a couple sections that could still be a little more clear I think.

File: notebooks/08B_modeling_parallel_sorting_data.ipynb (hunk 20 above)
Comment (michaelmckinsey1) [reply notebooks/08B_modeling_parallel_sorting_data.ipynb]: 2248c132 - rename feature from node presence to "Present" to match paper

73c70cd6 - update 3c text

Yes it's described in the paper as "Present"

File: notebooks/08A_composing_parallel_sorting_data.ipynb (hunk 19 above)
Comment (michaelmckinsey1) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: 98ca5cb9 - Added the generalized tree to explain what we're targeting here.

Comment (dyokelson) [reply notebooks/08A_composing_parallel_sorting_data.ipynb]: This looks great - much clearer.

**Comment by dyokelson** [comment]
LGTM

**Review by dyokelson — APPROVED** [review]

**Comment by michaelmckinsey1** [comment]
~Depends on #51~

Hosting data elsewhere

**Comment by michaelmckinsey1** [comment]
Ok this is working with docker. The only outstanding issue is that 08A depends on Thicket newer than v2024.1.0, specifically commit 73de0ce. I currently have it upgrading the Thicket version to develop in 08A, however the notebook needs to be restarted after upgrading the Thicket version to reset the kernel. This will be fixed once we can reference the minor release.

**Review by slabasan — APPROVED** [review]
